File: WinForms/VB/HtmlMainDemo/Modules/AlertControl/AlertControlModule.vb
Imports System
Imports System.Drawing
Imports System.Windows.Forms
Imports DevExpress.DXperience.Demos
Imports DevExpress.HTML.Demos.Helpers.Data
Imports DevExpress.Utils
Imports DevExpress.Utils.Html
Imports DevExpress.Utils.Svg
Imports DevExpress.XtraBars.Alerter

Namespace DevExpress.HTML.Demos

    Public Partial Class AlertControlModule
        Inherits TutorialControlBase

        Private previewTemplate As String

        Private previewStyles As String

        Public Sub New()
            InitializeComponent()
            AddHandler templateViewer.ActiveTemplateChanged, AddressOf TemplateViewer_ActiveTemplateChanged
            templateViewer.Init(alertControl1)
            alertControlProperties.Init(alertControl1)
            AddHandler alertControl1.HtmlElementMouseClick, AddressOf AlertControl1_HtmlElementMouseClick
            AddHandler alertControl1.BeforeFormShow, AddressOf AlertControl1_BeforeFormShow
            UpdatePreview()
            Dim previewTimer As Timer = New Timer()
            previewTimer.Interval = 200
            AddHandler previewTimer.Tick, Sub(s, e) UpdatePreview()
            previewTimer.Start()
        End Sub

        Protected Overrides Sub OnLookAndFeelChanged()
            MyBase.OnLookAndFeelChanged()
            UpdatePropertiesHeight()
        End Sub

        Protected Overrides Sub OnLoad(ByVal e As EventArgs)
            MyBase.OnLoad(e)
            UpdatePropertiesHeight()
        End Sub

        Private Sub UpdatePropertiesHeight()
            mainPanel.Rows(4).Height = Math.Min(alertControlProperties.GetBestHeight(), mainPanel.Height \ 2) / ScaleUtils.ScaleFactor.Height
        End Sub

        Private Sub UpdatePreview()
            htmlPreview.HtmlImages = alertControl1.HtmlImages
            Dim newContext As AlertInfo = GetCurrentAlertInfo()
            If TypeOf htmlPreview.DataContext Is AlertInfo Then
                UpdateAlertinfo(CType(htmlPreview.DataContext, AlertInfo))
            Else
                htmlPreview.DataContext = GetCurrentAlertInfo()
            End If

            Dim currentTemlpate As HtmlTemplate = GetCurrentTemplate()
            If Not Equals(currentTemlpate.Styles, previewStyles) OrElse Not Equals(currentTemlpate.Template, previewTemplate) Then
                previewStyles = currentTemlpate.Styles
                previewTemplate = currentTemlpate.Template
                htmlPreview.HtmlTemplate.Set($"<div class=""preview-h-container""><div class=""preview-v-container"" id=""previewRoot"">{previewTemplate}</div></div>", $".preview-h-container{{display:flex;flex-direction:row;align-items:center;justify-content: center;height: 100%;}}.preview-v-container{{display:flex;flex-direction:column;}}{previewStyles}")
            End If
        End Sub

        Private Sub TemplateViewer_ActiveTemplateChanged(ByVal sender As Object, ByVal e As EventArgs)
            InitAlertInfo(templateViewer.ActiveTemplate)
        End Sub

        Private Sub InitAlertInfo(ByVal currentTemplate As HtmlTemplate)
            If currentTemplate Is glowTemplate Then
                alertInfoProperties.Init("Sample Alert", "The color of a glowing effect around this alert box changes depending on the color of the ""Image"" icon", True)
            ElseIf currentTemplate Is toastTemplate Then
                alertInfoProperties.Init("Windows 11 Inspired Toast", "Use the selector at the top right corner to change a template. Click the ""Show an Alert"" button to display a notification.", False)
            ElseIf currentTemplate Is simpleTemplate Then
                alertInfoProperties.Init("Sample Alert", "Use the selector at the top right corner to change a template. Click the ""Show an Alert"" button to display a notification.", True)
            ElseIf currentTemplate Is darkTemplate Then
                alertInfoProperties.Init(Nothing, "Use the selector at the top right corner to change a template. Click the ""Show an Alert"" button to display a notification.", False)
            End If
        End Sub

        Private Sub AlertControl1_HtmlElementMouseClick(ByVal sender As Object, ByVal e As AlertHtmlElementMouseEventArgs)
            If Equals(e.ElementId, "closeButton") OrElse e.ParentHasId("closeButton") OrElse Equals(e.ElementId, "okButton") OrElse e.ParentHasId("okButton") Then
                e.HtmlPopup.Close()
            Else
                e.HtmlPopup.Pinned = Not e.HtmlPopup.Pinned
            End If
        End Sub

        Private Sub AlertControl1_BeforeFormShow(ByVal sender As Object, ByVal e As AlertFormEventArgs)
            e.HtmlPopup.HtmlTemplate.Assign(GetCurrentTemplate())
        End Sub

        Private Function GetCurrentSvgImageColor() As String
            Dim image As SvgImage = alertInfoProperties.SvgImage
            If alertInfoProperties.SvgImage Is Nothing Then Return Nothing
            Return GetCurrentSvgImageColor(image.Root)
        End Function

        Private Function GetCurrentSvgImageColor(ByVal element As SvgElement) As String
            If Not String.IsNullOrEmpty(element.Fill) Then Return element.StyleName
            For Each child As SvgElement In element.Elements
                Dim color As String = GetCurrentSvgImageColor(child)
                If Not Equals(color, Nothing) Then Return color
            Next

            Return Nothing
        End Function

        Private Function GetPatchedGlowTemplate(ByVal template As HtmlTemplate) As HtmlTemplate
            Dim colorName As String = GetCurrentSvgImageColor()
            If String.IsNullOrEmpty(colorName) Then Return template
            Dim styles As String = template.Styles.Replace("@Black", "@" & colorName)
            Return New HtmlTemplate(template.Template, styles)
        End Function

        Private Function GetPatchedTemplate(ByVal template As HtmlTemplate) As HtmlTemplate
            If Equals(template.Name, glowTemplate.Name) Then
                Return GetPatchedGlowTemplate(template)
            Else
                Return template
            End If
        End Function

        Private Function GetCurrentTemplate() As HtmlTemplate
            Return GetPatchedTemplate(templateViewer.ActiveTemplate)
        End Function

        Private Function GetCurrentAlertInfo() As AlertInfo
            Dim info As AlertInfo = New AlertInfo(Nothing, Nothing)
            alertInfoProperties.UpdateAlertInfo(info)
            Return info
        End Function

        Private Sub UpdateAlertinfo(ByVal info As AlertInfo)
            alertInfoProperties.UpdateAlertInfo(info)
        End Sub

        Private Sub ShowAlertButton_Click(ByVal sender As Object, ByVal e As EventArgs)
            alertControl1.Show(FindForm(), GetCurrentAlertInfo())
        End Sub
    End Class
End Namespace

File: WinForms/CS/HtmlMainDemo/Modules/AlertControl/AlertControlModule.cs
using System;
using System.Collections;
using System.Drawing;
using System.Windows.Forms;
using DevExpress.DXperience.Demos;
using DevExpress.HTML.Demos.Helpers.Data;
using DevExpress.LookAndFeel;
using DevExpress.Utils;
using DevExpress.Utils.Html;
using DevExpress.Utils.Html.ViewInfo;
using DevExpress.Utils.Svg;
using DevExpress.XtraBars.Alerter;
using DevExpress.XtraEditors;
using DevExpress.XtraGrid.Views.WinExplorer;

namespace DevExpress.HTML.Demos {
    public partial class AlertControlModule : TutorialControlBase {
        string previewTemplate;
        string previewStyles;
        public AlertControlModule() {
            InitializeComponent();
            templateViewer.ActiveTemplateChanged += TemplateViewer_ActiveTemplateChanged;
            templateViewer.Init(this.alertControl1);
            alertControlProperties.Init(alertControl1);
            alertControl1.HtmlElementMouseClick += AlertControl1_HtmlElementMouseClick;
            alertControl1.BeforeFormShow += AlertControl1_BeforeFormShow;
            UpdatePreview();
            Timer previewTimer = new Timer();
            previewTimer.Interval = 200;
            previewTimer.Tick += (s, e) => { UpdatePreview(); };
            previewTimer.Start();
        }

        protected override void OnLookAndFeelChanged() {
            base.OnLookAndFeelChanged();
            UpdatePropertiesHeight();
        }
        protected override void OnLoad(EventArgs e) {
            base.OnLoad(e);
            UpdatePropertiesHeight();
        }

        void UpdatePropertiesHeight() {
            mainPanel.Rows[4].Height = Math.Min(alertControlProperties.GetBestHeight(), mainPanel.Height / 2) / ScaleUtils.ScaleFactor.Height;
        }

        void UpdatePreview() {
            htmlPreview.HtmlImages = alertControl1.HtmlImages;
            AlertInfo newContext = GetCurrentAlertInfo();
            if(htmlPreview.DataContext is AlertInfo)
                UpdateAlertinfo((AlertInfo)htmlPreview.DataContext);
            else htmlPreview.DataContext = GetCurrentAlertInfo();
            HtmlTemplate currentTemlpate = GetCurrentTemplate();
            if(currentTemlpate.Styles != previewStyles || currentTemlpate.Template != previewTemplate) {
                previewStyles = currentTemlpate.Styles;
                previewTemplate = currentTemlpate.Template;
                htmlPreview.HtmlTemplate.Set($"<div class=\"preview-h-container\"><div class=\"preview-v-container\" id=\"previewRoot\">{previewTemplate}</div></div>",
                    $".preview-h-container{{display:flex;flex-direction:row;align-items:center;justify-content: center;height: 100%;}}.preview-v-container{{display:flex;flex-direction:column;}}{previewStyles}");
            }
        }

        private void TemplateViewer_ActiveTemplateChanged(object sender, EventArgs e) {
            InitAlertInfo(templateViewer.ActiveTemplate);
        }

        void InitAlertInfo(HtmlTemplate currentTemplate) {
            if(currentTemplate == glowTemplate)
                alertInfoProperties.Init("Sample Alert", "The color of a glowing effect around this alert box changes depending on the color of the \"Image\" icon", true);
            else if(currentTemplate == toastTemplate)
                alertInfoProperties.Init("Windows 11 Inspired Toast", "Use the selector at the top right corner to change a template. Click the \"Show an Alert\" button to display a notification.", false);
            else if(currentTemplate == simpleTemplate)
                alertInfoProperties.Init("Sample Alert", "Use the selector at the top right corner to change a template. Click the \"Show an Alert\" button to display a notification.", true);
            else if(currentTemplate == darkTemplate)
                alertInfoProperties.Init(null, "Use the selector at the top right corner to change a template. Click the \"Show an Alert\" button to display a notification.", false);
        }

        private void AlertControl1_HtmlElementMouseClick(object sender, AlertHtmlElementMouseEventArgs e) {
            if(e.ElementId == "closeButton" || e.ParentHasId("closeButton") ||
                e.ElementId == "okButton" || e.ParentHasId("okButton"))
                e.HtmlPopup.Close();
            else
                e.HtmlPopup.Pinned = !e.HtmlPopup.Pinned;
        }

        private void AlertControl1_BeforeFormShow(object sender, AlertFormEventArgs e) {
            e.HtmlPopup.HtmlTemplate.Assign(GetCurrentTemplate());
        }

        string GetCurrentSvgImageColor() {
            SvgImage image = alertInfoProperties.SvgImage;
            if(alertInfoProperties.SvgImage == null)
                return null;
            return GetCurrentSvgImageColor(image.Root);
        }
        string GetCurrentSvgImageColor(SvgElement element) {
            if(!String.IsNullOrEmpty(element.Fill))
                return element.StyleName;
            foreach(SvgElement child in element.Elements) {
                string color = GetCurrentSvgImageColor(child);
                if(color != null)
                    return color;
            }
            return null;
        }

        HtmlTemplate GetPatchedGlowTemplate(HtmlTemplate template) {
            string colorName = GetCurrentSvgImageColor();
            if(string.IsNullOrEmpty(colorName))
                return template;
            string styles = template.Styles.Replace("@Black", "@" + colorName);
            return new HtmlTemplate(template.Template, styles);
        }

        HtmlTemplate GetPatchedTemplate(HtmlTemplate template) {
            if(template.Name == glowTemplate.Name)
                return GetPatchedGlowTemplate(template);
            else return template;
        }

        HtmlTemplate GetCurrentTemplate() {
            return GetPatchedTemplate(templateViewer.ActiveTemplate);
        }
        AlertInfo GetCurrentAlertInfo() {
            AlertInfo info = new AlertInfo(null, null);
            alertInfoProperties.UpdateAlertInfo(info);
            return info;
        }

        void UpdateAlertinfo(AlertInfo info) {
            alertInfoProperties.UpdateAlertInfo(info);
        }

        private void ShowAlertButton_Click(object sender, EventArgs e) {
            alertControl1.Show(this.FindForm(), GetCurrentAlertInfo());
        }
    }
}
